pkg/integrations/v2/register.go

```diff
@@ -28,18 +34,48 @@ var (
 //
 // Register panics if cfg is not a pointer.
 func Register(cfg Config, ty Type) {
-	if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
-		panic(fmt.Sprintf("Register must be given a pointer, got %T", cfg))
+	registerIntegration(cfg, cfg.Name(), ty, nil)
+}
+
+func registerIntegration(v interface{}, name string, ty Type, upgrader UpgradeFunc) {
+	if reflect.TypeOf(v).Kind() != reflect.Ptr {
+		panic(fmt.Sprintf("Register must be given a pointer, got %T", v))
 	}
-	if _, exist := configFieldNames[reflect.TypeOf(cfg)]; exist {
-		panic(fmt.Sprintf("Integration %T registered twice", cfg))
+	if _, exist := integrationNames[name]; exist {
+		panic(fmt.Sprintf("Integration %q registered twice", name))
 	}
+	integrationNames[name] = struct{}{}
 
-	registeredIntegrations = append(registeredIntegrations, cfg)
+	registered = append(registered, v)
 
-	configTy := reflect.TypeOf(cfg)
+	configTy := reflect.TypeOf(v)
 	integrationTypes[configTy] = ty
-	configFieldNames[configTy] = cfg.Name()
+	configFieldNames[configTy] = name
+	upgraders[configTy] = upgrader
+}
+
+// RegisterLegacy registers a v1.Config. upgrader will be used to upgrade it.
+// upgrader will only be invoked after unmarshaling cfg from YAML, and the
+// upgraded Config will be unwrapped again when marshaling back to YAML.
+//
+// RegisterLegacy only exists for the transition period where the v2
+// integrations subsystem is an experiment. RegisterLegacy will be removed at a
+// later date.
+func RegisterLegacy(cfg v1.Config, ty Type, upgrader UpgradeFunc) {
+	registerIntegration(cfg, cfg.Name(), ty, upgrader)
+}
```

Comment:
This is much more concise, and makes the intent more clear to me. 🎉 

Leaving the logic for `registerIntegration` here *could* allow for other weird corner cases, should they arise, but you'd have to explicitly expose them like you're doing here.